File: go/internal/tools/storage.go
package tools

import (
	"encoding/json"
	"fmt"
	"sync"
)

// In-memory storage
var (
	storage      = make(map[string]string)
	storageMutex = &sync.RWMutex{}
)

// StorageSetInput represents input for storage_set
type StorageSetInput struct {
	Key   string `json:"key"`
	Value string `json:"value"`
}

// StorageGetInput represents input for storage_get
type StorageGetInput struct {
	Key string `json:"key"`
}

// StorageDeleteInput represents input for storage_delete
type StorageDeleteInput struct {
	Key string `json:"key"`
}

// StorageSet stores a key-value pair
func StorageSet(args map[string]interface{}) (interface{}, error) {
	argBytes, err := json.Marshal(args)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal arguments: %w", err)
	}

	var input StorageSetInput
	if err := json.Unmarshal(argBytes, &input); err != nil {
		return nil, fmt.Errorf("failed to unmarshal arguments: %w", err)
	}

	if input.Key == "" {
		return nil, fmt.Errorf("key is required")
	}

	storageMutex.Lock()
	defer storageMutex.Unlock()

	storage[input.Key] = input.Value

	return map[string]interface{}{
		"success": true,
		"message": fmt.Sprintf("Stored value for key '%s'", input.Key),
	}, nil
}

// StorageGet retrieves a value by key
func StorageGet(args map[string]interface{}) (interface{}, error) {
	argBytes, err := json.Marshal(args)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal arguments: %w", err)
	}

	var input StorageGetInput
	if err := json.Unmarshal(argBytes, &input); err != nil {
		return nil, fmt.Errorf("failed to unmarshal arguments: %w", err)
	}

	if input.Key == "" {
		return nil, fmt.Errorf("key is required")
	}

	storageMutex.RLock()
	defer storageMutex.RUnlock()

	value, exists := storage[input.Key]
	if !exists {
		return map[string]interface{}{
			"found": false,
			"key":   input.Key,
		}, nil
	}

	return map[string]interface{}{
		"found": true,
		"key":   input.Key,
		"value": value,
	}, nil
}

// StorageDelete deletes a key-value pair
func StorageDelete(args map[string]interface{}) (interface{}, error) {
	argBytes, err := json.Marshal(args)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal arguments: %w", err)
	}

	var input StorageDeleteInput
	if err := json.Unmarshal(argBytes, &input); err != nil {
		return nil, fmt.Errorf("failed to unmarshal arguments: %w", err)
	}

	if input.Key == "" {
		return nil, fmt.Errorf("key is required")
	}

	storageMutex.Lock()
	defer storageMutex.Unlock()

	_, exists := storage[input.Key]
	if !exists {
		return map[string]interface{}{
			"success": false,
			"message": fmt.Sprintf("Key '%s' not found", input.Key),
		}, nil
	}

	delete(storage, input.Key)

	return map[string]interface{}{
		"success": true,
		"message": fmt.Sprintf("Deleted key '%s'", input.Key),
	}, nil
}

// StorageList lists all stored keys
func StorageList(args map[string]interface{}) (interface{}, error) {
	storageMutex.RLock()
	defer storageMutex.RUnlock()

	keys := make([]string, 0, len(storage))
	for key := range storage {
		keys = append(keys, key)
	}

	return map[string]interface{}{
		"keys":  keys,
		"count": len(keys),
	}, nil
}

// GetStorageSetSchema returns the JSON schema for storage_set
func GetStorageSetSchema() map[string]interface{} {
	return map[string]interface{}{
		"type": "object",
		"properties": map[string]interface{}{
			"key": map[string]interface{}{
				"type":        "string",
				"description": "The key to store the value under",
			},
			"value": map[string]interface{}{
				"type":        "string",
				"description": "The value to store",
			},
		},
		"required": []string{"key", "value"},
	}
}

// GetStorageGetSchema returns the JSON schema for storage_get
func GetStorageGetSchema() map[string]interface{} {
	return map[string]interface{}{
		"type": "object",
		"properties": map[string]interface{}{
			"key": map[string]interface{}{
				"type":        "string",
				"description": "The key to retrieve",
			},
		},
		"required": []string{"key"},
	}
}

// GetStorageDeleteSchema returns the JSON schema for storage_delete
func GetStorageDeleteSchema() map[string]interface{} {
	return map[string]interface{}{
		"type": "object",
		"properties": map[string]interface{}{
			"key": map[string]interface{}{
				"type":        "string",
				"description": "The key to delete",
			},
		},
		"required": []string{"key"},
	}
}

// GetStorageListSchema returns the JSON schema for storage_list
func GetStorageListSchema() map[string]interface{} {
	return map[string]interface{}{
		"type":       "object",
		"properties": map[string]interface{}{},
	}
}

File: rust/mcp-core/src/tools/storage.rs
use super::ToolHandler;
use crate::types::Tool;
use anyhow::{anyhow, Result};
use serde::Deserialize;
use serde_json::{json, Value};
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;

lazy_static::lazy_static! {
    static ref STORAGE: Arc<RwLock<HashMap<String, String>>> = Arc::new(RwLock::new(HashMap::new()));
}

// Storage Set Tool
#[derive(Debug, Deserialize)]
struct StorageSetInput {
    key: String,
    value: String,
}

pub struct StorageSetTool;

#[async_trait::async_trait]
impl ToolHandler for StorageSetTool {
    fn definition(&self) -> Tool {
        Tool {
            name: "storage_set".to_string(),
            description: "Store a key-value pair in memory".to_string(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "key": {
                        "type": "string",
                        "description": "The key to store the value under"
                    },
                    "value": {
                        "type": "string",
                        "description": "The value to store"
                    }
                },
                "required": ["key", "value"]
            }),
        }
    }

    async fn execute(&self, args: HashMap<String, Value>) -> Result<Value> {
        let input: StorageSetInput = serde_json::from_value(json!(args))?;

        if input.key.is_empty() {
            return Err(anyhow!("Key is required"));
        }

        let mut storage = STORAGE.write().await;
        storage.insert(input.key.clone(), input.value);

        Ok(json!({
            "success": true,
            "message": format!("Stored value for key '{}'", input.key)
        }))
    }
}

// Storage Get Tool
#[derive(Debug, Deserialize)]
struct StorageGetInput {
    key: String,
}

pub struct StorageGetTool;

#[async_trait::async_trait]
impl ToolHandler for StorageGetTool {
    fn definition(&self) -> Tool {
        Tool {
            name: "storage_get".to_string(),
            description: "Retrieve a value by key from memory".to_string(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "key": {
                        "type": "string",
                        "description": "The key to retrieve"
                    }
                },
                "required": ["key"]
            }),
        }
    }

    async fn execute(&self, args: HashMap<String, Value>) -> Result<Value> {
        let input: StorageGetInput = serde_json::from_value(json!(args))?;

        if input.key.is_empty() {
            return Err(anyhow!("Key is required"));
        }

        let storage = STORAGE.read().await;

        match storage.get(&input.key) {
            Some(value) => Ok(json!({
                "found": true,
                "key": input.key,
                "value": value
            })),
            None => Ok(json!({
                "found": false,
                "key": input.key
            })),
        }
    }
}

// Storage Delete Tool
#[derive(Debug, Deserialize)]
struct StorageDeleteInput {
    key: String,
}

pub struct StorageDeleteTool;

#[async_trait::async_trait]
impl ToolHandler for StorageDeleteTool {
    fn definition(&self) -> Tool {
        Tool {
            name: "storage_delete".to_string(),
            description: "Delete a key-value pair from memory".to_string(),
            input_schema: json!({
                "type": "object",
                "properties": {
                    "key": {
                        "type": "string",
                        "description": "The key to delete"
                    }
                },
                "required": ["key"]
            }),
        }
    }

    async fn execute(&self, args: HashMap<String, Value>) -> Result<Value> {
        let input: StorageDeleteInput = serde_json::from_value(json!(args))?;

        if input.key.is_empty() {
            return Err(anyhow!("Key is required"));
        }

        let mut storage = STORAGE.write().await;

        match storage.remove(&input.key) {
            Some(_) => Ok(json!({
                "success": true,
                "message": format!("Deleted key '{}'", input.key)
            })),
            None => Ok(json!({
                "success": false,
                "message": format!("Key '{}' not found", input.key)
            })),
        }
    }
}

// Storage List Tool
pub struct StorageListTool;

#[async_trait::async_trait]
impl ToolHandler for StorageListTool {
    fn definition(&self) -> Tool {
        Tool {
            name: "storage_list".to_string(),
            description: "List all stored keys".to_string(),
            input_schema: json!({
                "type": "object",
                "properties": {}
            }),
        }
    }

    async fn execute(&self, _args: HashMap<String, Value>) -> Result<Value> {
        let storage = STORAGE.read().await;
        let keys: Vec<String> = storage.keys().cloned().collect();

        Ok(json!({
            "keys": keys,
            "count": keys.len()
        }))
    }
}
